neuraxle/metaopt/callbacks.py

```diff
@@ -0,0 +1,250 @@
+            validation_data_container=validation_data_container,
+            context=context
+        )
+
+        pipeline = trainer.refit(repo_trial.pipeline, data_container, context)
+
+
+    .. seealso::
+        :class:`AutoML`,
+        :class:`Trial`,
+        :class:`HyperparamsRepository`,
+        :class:`HyperparameterOptimizer`,
+        :class:`RandomSearchHyperparameterOptimizer`,
+        :class:`DataContainer`
+    """
+
+    def __init__(self, n_epochs_without_improvement, higher_score_is_better):
+        self.higher_score_is_better = higher_score_is_better
+        self.n_epochs_without_improvement = n_epochs_without_improvement
+
+    def call(
+            self,
+            trial: Trial,
+            epoch_number: int,
+            total_epochs: int,
+            input_train: DataContainer,
+            pred_train: DataContainer,
+            input_val: DataContainer,
+            pred_val: DataContainer,
+            is_finished_and_fitted: bool
+    ):
+        validation_scores = trial.get_validation_scores()
+        if len(validation_scores) > self.n_epochs_without_improvement:
+            if validation_scores[-self.n_epochs_without_improvement] >= validation_scores[-1] and self.higher_score_is_better:
+                return True
+            if validation_scores[-self.n_epochs_without_improvement] <= validation_scores[-1] and not self.higher_score_is_better:
+                return True
+        return False
+
+
+class MetaCallback(BaseCallback):
+    def __init__(self, wrapped_callback: BaseCallback):
+        self.wrapped_callback = wrapped_callback
+
+    @abstractmethod
+    def call(
+            self,
+            trial: Trial,
+            epoch_number: int,
+            total_epochs: int,
+            input_train: DataContainer,
+            pred_train: DataContainer,
+            input_val: DataContainer,
+            pred_val: DataContainer,
+            is_finished_and_fitted: bool
+    ):
+        pass
+
+class IfBestScore(MetaCallback):
+    def call(self, trial: Trial, epoch_number: int, total_epochs: int, input_train: DataContainer,
+             pred_train: DataContainer, input_val: DataContainer, pred_val: DataContainer, is_finished_and_fitted: bool):
+        if trial.is_new_best_score():
+            if self.wrapped_callback.call(
+                    trial,
+                    epoch_number,
+                    total_epochs,
+                    input_train,
+                    pred_train,
+                    input_val,
+                    pred_val,
+                    is_finished_and_fitted
+            ):
+                return True
+        return False
+
+class IfLastStep(MetaCallback):
+    def call(self, trial: Trial, epoch_number: int, total_epochs: int, input_train: DataContainer,
+             pred_train: DataContainer, input_val: DataContainer, pred_val: DataContainer, is_finished_and_fitted: bool):
+        if epoch_number == total_epochs - 1:
```

Comment:
or is_finished_and_fitted

Reply:
If this isn't added here, we need 2 more callbacks for the difference between finished and last. Because if we do early stopping we might not reach last.

Reply:
And also the negative versions of those callbacks as ifNot rather than if.